Clear PATH cache when mutating PATH

last-genius opened the pull request:
Fixes: https://github.com/oils-for-unix/oils/issues/2445

andychu reviewed (commented):
This looks nice and simple!  Thank you

One issue is that YSH uses ENV.PATH

```
$ ysh -c 'echo $[ENV.PATH]; setvar ENV.PATH = "z"; ls'
/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games:/home/andy/bin:/wedge/oils-for-unix.org/pkg/python2/2.7.18/bin/
  echo $[ENV.PATH]; setvar ENV.PATH = "z"; ls
                                           ^~
[ -c flag ]:1: Command 'ls' not found (OILS-ERR-100)
[ -c flag ]:1: errexit PID 2636547: Command failed with status 127
```

Actually I think a first step would be to add a failing test ...  the hook may be a bit trickier (?)

----

I think `spec/path-cache` can be folded into `spec/command_`, where other cases for the PATH cache are

And then maybe add an analogous file `spec/ysh-command` for the YSH path cache test

last-genius commented:
Added a test for YSH (and made it pass), and in the process discovered there's another way of doing PATH mutation in osh, so added another test and fixed that as well.

andychu commented on osh/cmd_eval.py:
```diff
@@ -1030,6 +1030,11 @@ def _DoMutation(self, node):
                                                 loc.Missing)
+                            # Hack: let SearchPath know it needs to reset the path
+                            # cache before looking up a command
+                            if key == 'PATH':
```
Hm let me think about this  ... it seems a bit overzealous,  since it also clears the cache on `setvar other.PATH = 42`, I think

We could define YSH so you do `hash -r` aka `hash --clear` or something ... although maybe that's too unfriendly

andychu commented:
I think it would be OK to land a fix for OSH first, and then have the failing spec test for YSH as a TODO

So we're not blocked on a design issue

Although I guess this one doesn't necessarily make any regtest/aports pass -- although it might!

andychu commented on core/state.py:
```diff
@@ -2023,6 +2030,12 @@ def SetNamed(self, lval, val, which_scopes, flags=0):
         can also be used to initialize?
         """
+
+        # Hack: let SearchPath know it needs to reset the path
```
I think the assumption here is that this is the ONLY way to change PATH

We also have an issue with `LANG=C` -- in bash, that doesn't just mutate a variable, it also makes a function call like `setlocale()` !!

andychu reviewed (commented):
Hm now I'm thinking about what we need for LANG=C ... hm

I guess for most of these, we tend to CACHE the parsing

https://github.com/oils-for-unix/oils/issues/527

e.g. GLOBIGNORE and PATH

We don't do anything special when you mutate them ... 

But LANG=C and PATH= are different!

andychu commented on core/state.py:
```diff
@@ -1991,6 +1993,11 @@ def SetNamedYsh(self, lval, val, which_scopes, flags=0):
 
+        # Hack: let SearchPath know it needs to reset the path
+        # cache before looking up a command
+        if lval.name == 'PATH':
```
OK good point, that is why I'm thinking about the invariants here

It is a fuzzy line between OSH and YSH

And I even wonder if we need something besides ENV.PATH in YSH ...
